### PR TITLE
fix(consensus_adapter): reduce shared object transaction amplification by ~3

### PR DESCRIPTION
**Pull request by huitseeker**
This submits shared object transactions to only (f+1) validators.

Context: Each validator submits all shared object transactions it knows of, creating a shared object throughput amplification of 3f+1 over the optimum,

The Issue: This throughput amplification creates a burden on all subsequent steps in the pipeline.

The Mitigation: submit deterministically to (f+1) validators by stake, ensuring at least one honest submission.

### CODE REVIEW

File: crates/sui-core/src/consensus_adapter.rs [hunk 1]
```diff
@@ -240,9 +245,38 @@ impl ConsensusAdapter {
...
+        let mut validators = committee.voting_rights.clone();
+        validators.shuffle(&mut rng);
+
+        // the last (f+1) elements by weight are the submitters for this transaction
+        let mut total_weight = 0u64;
+        let mut submitters = HashSet::new();
+        while total_weight < committee.validity_threshold() {
+            if let Some((name, weight)) = validators.pop() {
+                total_weight += weight;
+                submitters.insert(name);
```
Comment (lxfind) [inline crates/sui-core/src/consensus_adapter.rs:267]: Instead of adding the name to submitters, you could directly check if the name is self here.

**Review by lxfind — APPROVED** [review]

**Comment by huitseeker** [comment]
> What's the general trade-off between doing this within each validators vs doing this in the authority_aggregator?

1. in so far as:
    1. the `AuthorityAggregator` is used in various client contexts (e.g. by the benchmark)
    2. the transaction submission logic is indifferentiated across shared object TXs vs. the rest,
    Then, it should not be up to the client to tune any amplification in shared object throughput,
2. it's less important as this could be hacked around, but in so far as the `AuthorityAggregator` is used in non-consensus contexts, we'd have to include custom logic to figure out those few (shared object) transactions which require such a bespoke submission treatment. In the current PR we reuse the existing branching logic in `handle_certificate`.

**Review by andll — CHANGES_REQUESTED** [review]
I think that current consensus adapter needs more work then simply changing `should_submit` - the waiter correctness actually depends on the fact that everyone sends transaction

**Comment by andll** [comment]
To elaborate a bit more on what I think is the problem with this diff - 
Current logic in `ConsensusAdapter` is this:
- Register oneshot with waiter
- Submit transaction to consensus
- Wait for transaction to come from consensus
Current logic is always correct because we register waiter before submitting transaction, so it will always come back to unblock waiter

With this fix we break this assumption - we register oneshot with waiter but might not send a transaction, there is no guarantee that waiter will be unblocked(other then timeout)

**Comment by huitseeker** [comment]
A shared object transaction is only submitted after [checking it has not otherwise been executed](https://github.com/MystenLabs/sui/blob/bd0c07c5b11077149623f3fd4ed0c984131abfc4/crates/sui-core/src/authority_server.rs#L401-L418). 

Assume the current validator is not among the submitters for a given transaction T (i.e. `should_submit` returns false). There are at least (f+1) submitters (see unit test), among which at least one that will eventually proceed (modulo re-submissions), because its submitter is honest (see BFT assumptions). This honest submission is either in the past - meaning it has already been through the `ConsensusListener` - or the future.

- if the honest submission is in the past, then we have a contradiction: we should not find ourselves in this code path, because we should pick up that this transaction has been executed, so there is no waiter to "unblock",
- if the honest submission is in the future, it will lead our waiter to deliver.

**Comment by akichidis** [comment]
> if the honest submission is in the past, then we have a contradiction: we should not find ourselves in this code path, because we should pick up that this transaction has been executed, so there will be no waiter to "unblock",

@huitseeker do we have any guarantees that the `ConsensusAdapter` and `ConsensusListener` are working "synchronously"? I am looking on the code and it seems that is possible by the moment we register a new `ConsensusListenerMessage::New` message to the `ConsensusListener` , the transaction we are interested at might have already been sequenced (another validator was actually faster than us)  - so we'll miss getting notified about it and won't execute (unless there is some other path that also update us). It seems that the `ConsensusListener` is running on its own tokio task and working completely independently from the submission path. So even if we made any prior checks before submitting we might still end up "losing" the sequenced transaction. Not very familiar with the code path so please take everything I said with a (generous) pinch of salt :)

**Comment by andll** [comment]
Yeah, the general reasoning that @huitseeker provides is correct, have we executed everything in non-concurrent environment, but since we do different things in different tasks there are race conditions with the way check is currently implemented.

**Comment by huitseeker** [comment]
The `ConsensusListener` [only processes consensus output messages for transactions for which the shared locks have been set](https://github.com/MystenLabs/sui/blob/bd0c07c5b11077149623f3fd4ed0c984131abfc4/crates/sui-core/src/authority.rs#L2074-L2080). Given the source linked in my above comment, that same `ConsensusListener` only receives consensus input (waiter) messages for transactions for which the shared locks have *not* been set. The behavior of the shared locks is (we hope) what ensures synchronization.

For the `ConsensusListener` to miss a waiter on transaction T, we would need to observe the following sequence:
1. there are *no* shared locks on T, which means none of the (up to f+1) sequencing attempts from the network has succeeded yet,
2. we register a waiter for T, but for some reason, this waiter is not unpacked from the `ConsensusListener::rx_consensus_input` channel (e.g. there are *many* other waiters to unpack in that same channel),
3. we receive T, set shared locks on T, and send it to the `ConsensusListener::tx_consensus_output`,
4. for some reason, we unpack that transaction from the `ConsensusListener::rx_consensus_output` channel before we touch the waiter channel.

I would have a few remarks:

a. this can already happen: the only thing that changes w/o the present PR is the parenthesis "(up to f+1)" above that reverts to "(up to 3f+1)". The [above logic](https://github.com/MystenLabs/sui/pull/4979#issuecomment-1267522120) ("we register waiter before submitting transaction") also hinged on the same assumption of sequential execution,
b. any successful processing of T will set shared locks, and obviate any further submission,
c. the remediation path for this scenario is a timeout, then one re-attempt to handle the certificate, which will notice the shared locks, as long as by then the pending transaction has been unpacked from the `consensus_output` channel.
d. the rigorous way of solving this is to equip the `ConsensusListener` with a single channel, and make it receive `Either<ConsensusListenerMessage, Vec<u8>>`, leveraging the ability of `tokio::sync::mpsc` channels to deliver messages in-order.

I suggest we treat this edge case (using d) in a later PR.

**Comment by huitseeker** [comment]
> So the problem does not happen now, as you always wait after submitting txn and because we always register waiter before submitting request there is no room for race conditions there.

If I understand your reasoning correctly, you're telling me that for a given message, this:
https://github.com/MystenLabs/sui/blob/0f346b7977c34efe4604185f08931734e874f9ba/crates/sui-core/src/consensus_adapter.rs#L414-L434

Always fires ahead of that:
https://github.com/MystenLabs/sui/blob/0f346b7977c34efe4604185f08931734e874f9ba/crates/sui-core/src/consensus_adapter.rs#L291-L313

Is that correct?

**Comment by huitseeker** [comment]
The race condition outlined [above](https://github.com/MystenLabs/sui/pull/4979#issuecomment-1268772320) was fixed (see 'd.') on Oct. 7 in #5038 (thanks @andll !). This should therefore be ready to go.

**Comment by andll** [comment]
I don't think single channel fixed the race condition. The race is because we check the database first, then registering consensus waiter, so between those two operations something can get in the database and notify (not yet registered) waiters.

**Comment by mwtian** [comment]
As mentioned by @lxfind earlier, I feel it seems more natural to select authorities to submit certificate inside `AuthorityAggregator`. The malicious client problem can be addressed with rate limiting, which we will need to implement anyway.

Since byzantine authority can only slow down submissions or drop certificates, I think we can just reuse the retry idiom from a few other places, where a client submits a certificate to a random authority weighted by stake, waits for 5~10s to see if the certificate is in Narwhal output. If not, the client re-submits the certificate at a different random authority. In normal cases, we will just have 1 submission to narwhal per certificate. If there is a byzantine authority, it will cause latency issues but as long as we can detect and mitigate that it shouldn't be a big problem?

**Comment by huitseeker** [comment]
> As mentioned by @lxfind earlier, I feel it seems more natural to select authorities to submit certificate inside AuthorityAggregator. The malicious client problem can be addressed with rate limiting, which we will need to implement anyway.

How would you defend against a Sybil attack?

**Review by andll — APPROVED** [review]
I think with https://github.com/MystenLabs/sui/pull/5986 merged we are good to go with this change. Thanks for the patience @huitseeker

**Pull request merged**